src/tool_parsecfg.c

```diff
@@ -335,39 +343,23 @@ static const char *unslashquote(const char *line, char *param)
 
 /*
  * Reads a line from the given file, ensuring is NUL terminated.
- * The pointer must be freed by the caller.
- * NULL is returned on an out of memory condition.
+ * FALSE is returned on an out of memory condition.
  */
-static char *my_get_line(FILE *fp)
+static bool my_get_line(FILE *fp, struct curlx_dynbuf *db)
 {
   char buf[4096];
-  char *nl = NULL;
-  char *line = NULL;
-
   do {
-    if(NULL == fgets(buf, sizeof(buf), fp))
-      break;
-    if(!line) {
-      line = strdup(buf);
-      if(!line)
-        return NULL;
-    }
-    else {
-      char *ptr;
-      size_t linelen = strlen(line);
-      ptr = realloc(line, linelen + strlen(buf) + 1);
-      if(!ptr) {
-        Curl_safefree(line);
-        return NULL;
-      }
-      line = ptr;
-      strcpy(&line[linelen], buf);
+    /* fgets() returns s on success, and NULL on error or when end of file
+       occurs while no characters have been read. */
+    if(!fgets(buf, sizeof(buf), fp)) {
+      if(feof(fp))
```

Comment:
Here's a way I think we can do it, this way it only returns true on eof if some data was read. This is a slight change in behavior because it won't return true for an empty file anymore.
~~~diff
diff --git a/src/tool_parsecfg.c b/src/tool_parsecfg.c
index 46b4df9..ba22859 100644
--- a/src/tool_parsecfg.c
+++ b/src/tool_parsecfg.c
@@ -348,11 +348,12 @@ static const char *unslashquote(const char *line, char *param)
 static bool my_get_line(FILE *fp, struct curlx_dynbuf *db)
 {
   char buf[4096];
+  size_t oldlen = curlx_dyn_len(db);
   do {
     /* fgets() returns s on success, and NULL on error or when end of file
        occurs while no characters have been read. */
     if(!fgets(buf, sizeof(buf), fp)) {
-      if(feof(fp))
+      if(feof(fp) && oldlen != curlx_dyn_len(db))
         break;
       return FALSE;
     }
~~~
I can't figure out how to span multiple lines using the suggestion feature

Reply:
Reading an empty file should not cause am error. What case does this addition help with?

Reply:
If we return true on eof when no data has been read then it's just going to loop forever because parsecfg won't know that there are no more lines. Try running it now. We have to be able to signal to parsecfg that the end of the file has been reached and no more lines are available (return false). I disagree with your position on an empty file, as far as I'm concerned it has 0 lines.

before:
printf "foo\n" > curlcfg # two lines
printf "foo" > curlcfg # one line
printf "" > curlcfg # one line

after:
printf "foo\n" > curlcfg # one line
printf "foo" > curlcfg # one line
printf "" > curlcfg # no lines

Reply:
Ah right. The function returns if the loop should continue or not, so we're back to the simpler:
~~~c
{
  char buf[4096];
  do {
    /* fgets() returns s on success, and NULL on error or when end of file
       occurs while no characters have been read. */
    if(!fgets(buf, sizeof(buf), fp) ||
       curlx_dyn_add(db, buf))
      /* failure */
      return FALSE;
  } while(!strchr(buf, '\n'));

  return TRUE; /* suceess */
}

Reply:
This matches what the function in current master does.